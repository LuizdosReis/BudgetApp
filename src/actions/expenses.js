import uuid from 'uuid';

export const add = ({
  description = '', category = null, createAt = 0, amount = 0,
} = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    category,
    createAt,
    amount,
  },
});

export const remove = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id,
});

export const edit = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
});
