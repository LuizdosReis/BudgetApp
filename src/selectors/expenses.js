import moment from 'moment';

export default (expenses, {
  startDate, endDate, description, sortBy,
}) => expenses
  .filter((expense) => {
    const createAtMoment = moment(expense.createAt);
    const startDateMatch = startDate ? createAtMoment.isSameOrAfter(startDate, 'day') : true;
    const endDateMatch = endDate ? createAtMoment.isSameOrBefore(endDate, 'day') : true;
    const descriptionMatch = description
      ? expense.description.toLowerCase().includes(description.toLowerCase())
      : true;
    return startDateMatch && endDateMatch && descriptionMatch;
  })
  .sort((a, b) => {
    if (sortBy === 'description') {
      return a.description < b.description ? -1 : 1;
    }
    if (sortBy === 'amount') {
      return a.amount < b.amount ? -1 : 1;
    }
    return a.createAt < b.createAt ? -1 : 1;
  });
