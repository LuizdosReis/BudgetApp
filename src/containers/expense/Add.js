import React from 'react';
import { connect } from 'react-redux';

import { add } from '../../actions/expenses';
import Form from './Form';
import Header from '../../pages/Header';

export class Add extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/expenses');
  };

  render() {
    return (
      <div id="layout">
        <Header title="Adicione uma despesa" />
        <main className="content">
          <Form onSubmit={this.onSubmit} />
        </main>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(add(expense)),
});

export default connect(
  undefined,
  mapDispatchToProps,
)(Add);
