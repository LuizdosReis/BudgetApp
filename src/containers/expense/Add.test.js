import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Add } from './Add';
import expenses from '../../fixtures/expenses';

let addExpenseSpy;
let historySpy;
let wrapper;

beforeEach(() => {
  addExpenseSpy = jest.fn();
  historySpy = { push: jest.fn() };
  wrapper = shallow(<Add add={addExpenseSpy} history={historySpy} />);
});

test('should render Add correctly', () => {
  expect(toJson(wrapper)).toMatchSnapshot();
});

test('should handle submit', () => {
  wrapper.find('Form').prop('onSubmit')(expenses[0]);
  expect(addExpenseSpy).toHaveBeenLastCalledWith(expenses[0]);
  expect(historySpy.push).toHaveBeenLastCalledWith('/expenses');
});
