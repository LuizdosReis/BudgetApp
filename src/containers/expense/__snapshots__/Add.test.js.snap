// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Add correctly 1`] = `
<div
  id="layout"
>
  <Header
    title="Adicione uma despesa"
  />
  <main
    className="content"
  >
    <Form
      onSubmit={[Function]}
    />
  </main>
</div>
`;
